### PR TITLE
TLS 1.3 server-side mutual authentication implementation

### PR DESCRIPTION
**Pull request by ryru**
This PR covers:

- #3

### CODE REVIEW

File: src/libtls/tls_server.c [hunk 1]
```diff
@@ -1589,6 +1587,12 @@ METHOD(tls_handshake_t, build, status_t,
 			case STATE_CERT_VERIFY_SENT:
 				return send_finished(this, type, writer);
 			case STATE_FINISHED_SENT:
+				if (!this->crypto->derive_app_keys(this->crypto, FALSE))
+				{
+					this->alert->add(this->alert, TLS_FATAL, TLS_INTERNAL_ERROR);
+					return NEED_MORE;
+				}
+				this->crypto->change_cipher(this->crypto, FALSE);
```
Comment (tobiasbrunner) [inline src/libtls/tls_server.c:1688]: Wouldn't it be enough to just move this call here and do the key derivation as before with a single call?

File: src/libtls/tls_crypto.c [hunk 2]
```diff
@@ -2132,16 +2159,36 @@ static bool derive_labeled_key(private_tls_crypto_t *this, bool server,
...
-									 NULL) ||
-		!this->hkdf->derive_key(this->hkdf, server,
-								aead->get_encr_key_size(aead), &key) ||
-		!this->hkdf->derive_iv(this->hkdf, server,
-							   aead->get_iv_size(aead), &iv))
+	if (label == TLS_HKDF_C_AP_TRAFFIC || label == TLS_HKDF_S_AP_TRAFFIC)
 	{
-		DBG1(DBG_TLS, "deriving key material failed");
-		goto out;
+		get_handshake_until(TLS_FINISHED, this->handshake, &hs);
```
Comment (tobiasbrunner) [inline src/libtls/tls_crypto.c]: Shouldn't it be possible to avoid this by doing the key derivation at the right time?

File: src/libtls/tls_peer.c [hunk 3]
```diff
@@ -603,30 +603,7 @@ static status_t process_certificate(private_tls_peer_t *this,
...
-			{
-				public = current->get_ref(current);
-				this->server_auth->merge(this->server_auth, auth, FALSE);
-				break;
-			}
-		}
-		enumerator->destroy(enumerator);
-	}
-	return public;
+	return tls_find_public_key(this->server_auth);
```
Comment (tobiasbrunner) [inline src/libtls/tls_peer.c]: Why not replace `find_public_key()` completely?

File: src/libtls/tests/suites/test_socket.c [hunk 4]
```diff
@@ -744,11 +738,6 @@ Suite *socket_suite_create()
 	add_tls_test(test_tls13, TLS_1_3);
 	suite_add_tcase(s, tc);
 
-	tc = tcase_create("TLS 1.3/mutl");
-	tcase_add_checked_fixture(tc, setup_creds, teardown_creds);
-	add_tls_test(test_tls13_mutual, TLS_1_3);
-	suite_add_tcase(s, tc);
-
```
Comment (tobiasbrunner) [inline src/libtls/tests/suites/test_socket.c:751]: Why did you remove this? (Interesting that it worked before, I guess the test does not actually enforce any mutual authentication.)

File: src/libtls/tls_server.c [hunk 5]
```diff
@@ -859,6 +879,14 @@ static status_t process_key_exchange(private_tls_server_t *this,
...
 }
 
+/**
+ * Find a trusted public key to encrypt/verify key exchange data
+ */
+static public_key_t *find_public_key(private_tls_server_t *this)
+{
+	return tls_find_public_key(this->peer_auth);
+}
+
```
Comment (tobiasbrunner) [inline src/libtls/tls_server.c]: Same as on the client, why not call `tls_find_public_key()` directly?

File: src/libtls/tls_server.c [hunk 6]
```diff
@@ -870,43 +898,70 @@ static status_t process_cert_verify(private_tls_server_t *this,
...
-									KEY_ANY, this->peer, this->peer_auth, TRUE);
-	while (enumerator->enumerate(enumerator, &public, &auth))
+	if (this->tls->get_version_max(this->tls) < TLS_1_3)
 	{
-		sig = bio_reader_create(reader->peek(reader));
-		verified = this->crypto->verify_handshake(this->crypto, public, sig);
-		sig->destroy(sig);
-		if (verified)
+		enumerator = lib->credmgr->create_public_enumerator(lib->credmgr,
+								KEY_ANY, this->peer, this->peer_auth, TRUE);
```
Comment (tobiasbrunner) [inline src/libtls/tls_server.c]: I think you could just use the same function that does the lookup via certificate for all TLS versions instead of looping through all keys here (not sure if that was ever necessary).

File: src/libtls/tls_server.c [hunk 7]
```diff
@@ -1359,6 +1413,55 @@ static status_t send_certificate_request(private_tls_server_t *this,
...
+
+		write_certificate_authorities(writer);
+	}
+	else
+	{
+		/* certificate request context as described in RFC 8446, section 4.3.2 */
+		writer->write_uint8(writer, 0);
+
+		extensions = bio_writer_create(32);
+		DBG2(DBG_TLS, "sending extension within certificate_reqst message: %N",
```
Comment (tobiasbrunner) [inline src/libtls/tls_server.c]: `certificate_reqst` looks a bit weird. Not sure if a message is even necessary.

File: src/libtls/tls_server.c [hunk 8]
```diff
@@ -1581,6 +1684,11 @@ METHOD(tls_handshake_t, build, status_t,
 			case STATE_CIPHERSPEC_CHANGED_OUT:
 				return send_encrypted_extensions(this, type, writer);
 			case STATE_ENCRYPTED_EXTENSIONS_SENT:
+				if (this->peer)
```
Comment (tobiasbrunner) [inline src/libtls/tls_server.c:1670]: This doesn't seem to be the case for older TLS versions.

File: src/libtls/tls_server.c [hunk 9]
```diff
@@ -1581,6 +1684,11 @@ METHOD(tls_handshake_t, build, status_t,
...
+				{
+					return send_certificate_request(this, type, writer);
+				}
+			case STATE_CERTREQ_SENT:
```
Comment (tobiasbrunner) [inline src/libtls/tls_server.c]: Is this fall-through on purpose? If so, add a comment to clarify.

File: scripts/tls_test.c [hunk 10]
```diff
@@ -402,7 +430,15 @@ int main(int argc, char *argv[])
 	cache = tls_cache_create(100, 30);
 	if (listen)
 	{
-		res = serve(host, server, times, cache, min_version, max_version);
+		if (cauth)
+		{
+			client = find_client_id();
```
Comment (tobiasbrunner) [inline scripts/tls_test.c]: Not sure if this makes sense as the client certificate won't be loaded on the server, i.e. that function will only find the server identity. Maybe just assign ID_ANY (`identification_create_from_encoding(ID_ANY, chunk_empty);`).  You also wouldn't need the `cauth` flag. Instead assign that identity to `client`, which you initialize to `NULL`, above in the arg parse loop.

**Review by tobiasbrunner — CHANGES_REQUESTED** [review]

File: src/libtls/tls_server.c (hunk 1 above)
Comment (ryru) [reply src/libtls/tls_server.c:1688]: In my understanding no. `derive_labeled_keys()` previously derived inbound and outbound keys and also set it to in both directions `this->aead_in` and `this->aead_out`.
It may be possible to combine `this->crypto->derive_app_keys()` and `this->crypto->change_cipher()` to a single call though.

File: src/libtls/tls_crypto.c (hunk 2 above)
Comment (ryru) [reply src/libtls/tls_crypto.c]: No, I think in the current code not. Key derivation also sets keys on `this->aead_in` respectively `this->aead_out` in `derive_labled_key()`:

```
	if (!aead->set_keys(aead, chunk_empty, key, iv))
	{
		DBG1(DBG_TLS, "setting AEAD key material failed");
		goto out;
	}
	success = TRUE;
```

File: src/libtls/tests/suites/test_socket.c (hunk 4 above)
Comment (ryru) [reply src/libtls/tests/suites/test_socket.c:751]: Until now the server never sent a _CertificateRequest_ message in TLS 1.3 test cases. With the server implementation it does now but the client is not ready now. I add the test again once the client is implemented.

File: src/libtls/tls_server.c (hunk 6 above)
Comment (ryru) [reply src/libtls/tls_server.c]: I add an implementation suggestion for this.

File: src/libtls/tls_server.c (hunk 7 above)
Comment (ryru) [reply src/libtls/tls_server.c]: I shortened the message to `sending CertificateRequest extension: %N` this is also more similar to the extension DBG2 messages sending in the ServerHello message. Do you agree?

File: src/libtls/tls_server.c (hunk 8 above)
Comment (ryru) [reply src/libtls/tls_server.c:1670]: Right, this only triggers a CertificateRequest message in TLS 1.3 but not in older TLS versions. Do I understand you correctly that such a functionality is desirable also for the TLS legacy branch?

File: scripts/tls_test.c (hunk 10 above)
Comment (ryru) [reply scripts/tls_test.c]: Thanks, this looks better.

File: src/libtls/tls_server.c [hunk 11]
```diff
@@ -879,46 +879,27 @@ static status_t process_key_exchange(private_tls_server_t *this,
...
+		return NEED_MORE;
+	}
 
-		if (!verified)
+	msg = reader->peek(reader);
+	if (this->tls->get_version_max(this->tls) < TLS_1_3)
+	{
+		if (!this->crypto->verify_handshake(this->crypto, public, reader))
 		{
 			DBG1(DBG_TLS, "no trusted certificate found for '%Y' to verify TLS peer",
```
Comment (tobiasbrunner) [inline src/libtls/tls_server.c]: This error message is not correct anymore. Should probably be the same as below for TLS 1.3. The alert should also be `TLS_DECRYPT_ERROR` in both TLS versions (see section 4.4.3). And there is a `public->destroy(public);` call required there before returning here. But shouldn't it be possible to combine the two paths?

File: src/libtls/tls_server.c (hunk 1 above)
Comment (tobiasbrunner) [reply src/libtls/tls_server.c:1688]: You're right. The keys in the existing `tls_aead_t` objects were set, which immediately changed the keys that are in use (i.e. calling `change_cipher()` is redundant after the first call). I guess this separation was necessary for the change_cipher_spec use case and the keys were also never changed in older versions. So for TLS 1.3, we could theoretically just call `change_cipher()` in `derive_handshake_keys()` and be done with it.

However, it might be cleaner to actually separate the key derivation from enabling a certain set of keys on the TLS protection layer. I guess the ownership of the `tls_aead_t` object would then be transferred to the protection layer whenever the method is called. New objects would later be created when a new set of keys is derived. That way the key derivation and the switch to new keys could both be done when appropriate.

File: src/libtls/tests/suites/test_socket.c (hunk 4 above)
Comment (tobiasbrunner) [reply src/libtls/tests/suites/test_socket.c:751]: I see.

File: src/libtls/tls_server.c (hunk 7 above)
Comment (tobiasbrunner) [reply src/libtls/tls_server.c]: I'd make it the same as the others, i.e. remove `CertificateRequest`, because we see `sending TLS CertificateRequest handshake` in the log right afterwards (well, after the DNs for the CAs - adding them before the other extensions might clarify this even further).

File: src/libtls/tls_server.c (hunk 8 above)
Comment (tobiasbrunner) [reply src/libtls/tls_server.c:1670]: That's what I'm wondering. It seems a certificate request was always sent with older versions, no matter if a peer identity was configured (which, according to the comment, was intended to enable this - if none is set, the client's authentication and certificate is verified but is allowed to fail, not sure what the purpose of that would be, though). So we perhaps should not make this conditional here either, or also add such a check for older versions. It really depends on whether it still works for our use cases if it's made conditional.

File: src/libtls/tls_server.c [hunk 12]
```diff
@@ -879,46 +879,27 @@ static status_t process_key_exchange(private_tls_server_t *this,
...
-			{
-				this->peer_auth->merge(this->peer_auth, auth, FALSE);
-				break;
-			}
-			DBG1(DBG_TLS, "signature verification failed, trying another key");
-		}
-		enumerator->destroy(enumerator);
+		DBG1(DBG_TLS, "no TLS public key found for peer '%Y'", this->peer);
+		this->alert->add(this->alert, TLS_FATAL, TLS_CERTIFICATE_UNKNOWN);
+		return NEED_MORE;
```
Comment (tobiasbrunner) [inline src/libtls/tls_server.c]: The `peer_auth_optional` use case must also be covered here.

**Review by tobiasbrunner — CHANGES_REQUESTED** [review]

File: src/libtls/tls_server.c (hunk 11 above)
Comment (ryru) [reply src/libtls/tls_server.c]: I addressed your comments.

File: src/libtls/tls_server.c (hunk 8 above)
Comment (ryru) [reply src/libtls/tls_server.c:1670]: Yes, this is the case. In older TLS versions the server always sent a _CertificateRequest_ message. I see three options:

a) I make it conditional in legacy TLS: This results in the situation that we currently not know, if it has an impact on how libtls is used in strongSwan.
b) I remove the conditional in TLS 1.3: This results in changes in the `tls_peer.c` because it receives a _CertificateRequest_ message but has not yet an implementation for mutual authentication.
c) leave as is: This does not break the old implementation an does not touch `tls_peer.c` in this commit.

I would suggest to leave it as is for now. Once the TLS 1.3 mutual authentication is done, I remove the conditional in a separate commit to get consistency. Do you agree?

File: src/libtls/tls_server.c (hunk 12 above)
Comment (ryru) [reply src/libtls/tls_server.c]: I think I addressed this. I do not see the purpose of `peer_auth_optional` though.

File: src/libtls/tls_server.c (hunk 1 above)
Comment (ryru) [reply src/libtls/tls_server.c:1688]: Let's discuss in today's meeting.

File: scripts/tls_test.c [hunk 13]
```diff
@@ -305,6 +325,7 @@ int main(int argc, char *argv[])
 			{"port",		required_argument,		NULL,		'p' },
 			{"cert",		required_argument,		NULL,		'x' },
 			{"key",			required_argument,		NULL,		'k' },
+			{"cafile",		required_argument,		NULL,		'f' },
```
Comment (tobiasbrunner) [inline scripts/tls_test.c]: Better call this `cacert`.

File: src/libtls/tls_hkdf.c [hunk 14]
```diff
@@ -67,11 +67,21 @@ struct private_tls_hkdf_t {
...
+	 * Current server to client traffic secret.
+	 */
 	chunk_t server_traffic_secret;
+
+	/**
+	 * Current implementation needs a copy of derived secrets to calculate the
+	 * proper finished key.
+	 */
+	chunk_t client_base_traffic_secret;
+	chunk_t server_base_traffic_secret;
```
Comment (tobiasbrunner) [inline src/libtls/tls_hkdf.c]: The names seem not ideal as there are different base keys throughout. Maybe call them what they actually are, `*_handshake_traffic_secret`.

Also, it might be preferable to store these a bit differently, so they can be passed around (also see below). For instance, define something like `struct cached_secrets_t { chunk_t client; chunk_t server; }` so you can then define these as `cached_secrets_t handshake_traffic_secrets;` and `cached_secrets_t traffic_secrets;`.

File: src/libtls/tls_hkdf.c [hunk 15]
```diff
@@ -482,6 +496,11 @@ static bool get_shared_label_keys(private_tls_hkdf_t *this, chunk_t label,
 
 	secret = server ? this->server_traffic_secret
 					: this->client_traffic_secret;
+	if (chunk_equals(chunk_from_str("finished"), label))
+	{
+		secret = server ? this->server_base_traffic_secret
+						: this->client_base_traffic_secret;
+	}
```
Comment (tobiasbrunner) [inline src/libtls/tls_hkdf.c]: Instead of this, pass a `cached_secrets_t*` to this helper (i.e. `&this->handshake_traffic_secrets` from `derive_finished`, `&this->traffic_secrets` from the others).

File: src/libtls/tls_hkdf.c [hunk 16]
```diff
@@ -446,12 +456,16 @@ METHOD(tls_hkdf_t, generate_secret, bool,
 	switch (label)
 	{
 		case TLS_HKDF_C_HS_TRAFFIC:
+			chunk_clear(&this->client_base_traffic_secret);
+			this->client_base_traffic_secret = chunk_clone(okm);
```
Comment (tobiasbrunner) [inline src/libtls/tls_hkdf.c]: Add a comment after this to clarify the fall-through (same below).

File: src/libtls/tls_crypto.c [hunk 17]
```diff
@@ -2164,8 +2164,17 @@ static bool derive_labeled_keys(private_tls_crypto_t *this,
 								tls_hkdf_label_t client_label,
 								tls_hkdf_label_t server_label)
 {
-	tls_aead_t *aead_c = this->aead_out, *aead_s = this->aead_in;
+	tls_aead_t *aead_c, *aead_s;
```
Comment (tobiasbrunner) [inline src/libtls/tls_crypto.c:2194]: Attaching this here, but it's really about the commit message. It describes that keys can be derived separately, but that's not the case. It's more that it allows separating key derivation from key switching if the derivation does not modify the  `tls_aead_t` objects currently in use by the protection layer.

File: src/libtls/tls_server.c [hunk 18]
```diff
@@ -1561,7 +1550,10 @@ METHOD(tls_handshake_t, build, status_t,
 				}
 				/* otherwise fall through to next state */
 			case STATE_KEY_EXCHANGE_SENT:
-				return send_certificate_request(this, type, writer);
+				if (this->peer)
+				{
+					return send_certificate_request(this, type, writer);
+				}
```
Comment (tobiasbrunner) [inline src/libtls/tls_server.c:1648]: Add a comment regarding the fall-through.

File: src/libtls/tls_server.c [hunk 19]
```diff
@@ -86,11 +86,6 @@ struct private_tls_server_t {
 	 */
 	identification_t *peer;
 
-	/**
-	 * Is it acceptable if we couldn't verify the peer certificate?
-	 */
-	bool peer_auth_optional;
```
Comment (tobiasbrunner) [inline src/libtls/tls_server.c:92]: In this patch, there should be additional removed code that used this flag (in `process_cert_verify()`). Maybe try editing this commit via `rebase -i`. You'll get some compile errors if you try to build this commit, so fix the code and amend the commit. There will then probably be some conflicts with the mutual auth commit when continuing the rebase, but you can  just apply all changes there as it does not use the flag anymore.

File: src/libtls/tls_server.c [hunk 20]
```diff
@@ -671,6 +671,12 @@ static status_t process_certificate(private_tls_server_t *this,
 		return NEED_MORE;
 	}
 	certs = bio_reader_create(data);
+	if (!certs->remaining(certs))
+	{
+		DBG1(DBG_TLS, "no cert sent by peer");
```
Comment (tobiasbrunner) [inline src/libtls/tls_server.c]: Better write `certificate` instead of `cert`. But I wonder whether this is actually necessary. If the client did not send a certificate, shouldn't `tls_find_public_key()` fail? (It requires a certificate in peer_auth.)

File: src/libtls/tls_server.c [hunk 21]
```diff
@@ -940,14 +940,7 @@ static status_t process_finished(private_tls_server_t *this,
 			this->alert->add(this->alert, TLS_FATAL, TLS_INTERNAL_ERROR);
 			return NEED_MORE;
 		}
-
-		if (!this->crypto->derive_app_keys(this->crypto))
```
Comment (tobiasbrunner) [inline src/libtls/tls_server.c:949]: I'd add some notes on the reasoning behind moving this code to the commit message. Because the side effect the additional messages sent by the client have on it might not be immediately obvious.

File: src/libtls/tls_server.c [hunk 22]
```diff
@@ -1580,12 +1635,25 @@ METHOD(tls_handshake_t, build, status_t,
...
 				return send_certificate_verify(this, type, writer);
 			case STATE_CERT_VERIFY_SENT:
 				return send_finished(this, type, writer);
 			case STATE_FINISHED_SENT:
+				if (!this->crypto->derive_app_keys(this->crypto))
+				{
+					this->alert->add(this->alert, TLS_FATAL, TLS_INTERNAL_ERROR);
+					return NEED_MORE;
+				}
+				this->crypto->change_cipher(this->crypto, FALSE);
```
Comment (tobiasbrunner) [inline src/libtls/tls_server.c:1688]: Maybe add a comment that explains that the inbound key is switched after processing the client's Finished message.

**Review by tobiasbrunner — CHANGES_REQUESTED** [review]

File: src/libtls/tls_server.c (hunk 20 above)
Comment (ryru) [reply src/libtls/tls_server.c]: If the client does not have any certificate to send it sends a _Certificate_ message without any certificate in it (certificate length == 0). In TLS 1.3 the connection is terminated by the server when no _CertificateVerify_ message is present.

In legacy TLS `process_cert_verify()`, which calls `tls_find_public_key()`, is never called because a client sends a _ChangeCipherSpec_ message between its _KeyExchange_ and _Finished_ messages. tls_fragmentation.c sets status to `FAILED` in the `process()` method since the `cipherspec_changed()` evaluates FALSE because `this->peer` is not null and the current state is `STATE_KEY_EXCHANGE_RECEIVED` and not `STATE_CERT_VERIFY_RECEIVED`. This means if a TLS 1.2 client tries to open a connection without a client certificate the handshake hangs after server output `processing TLS ChangeCipherSpec record (1 bytes)` which does not look right to me. The above change terminates the connection to the client.

Comment (ryru) [reply src/libtls/tls_server.c]: As discussed in today's meeting, we leave the current code since it is functional but open a new issue to investigate the case that the connection hangs after a _ChangeCipherSpec_ message.

File: src/libtls/tls_hkdf.c [hunk 23]
```diff
@@ -67,11 +72,14 @@ struct private_tls_hkdf_t {
...
-	 * Current implementation needs a copy of derived secrets to calculate the
-	 * proper finished key.
+	 * Handshake traffic secrets.
 	 */
-	chunk_t client_traffic_secret;
-	chunk_t server_traffic_secret;
+	cached_secrets_t handshake_traffic_secrets;
+
+	/**
+	 * Application traffic secrets.
```
Comment (tobiasbrunner) [inline src/libtls/tls_hkdf.c]: Not exclusively used for the application traffic secrets, initially also stores a copy of the handshake traffic secrets. So maybe use "Current" instead of "Application".

File: src/libtls/tls_hkdf.c [hunk 24]
```diff
@@ -446,16 +454,22 @@ METHOD(tls_hkdf_t, generate_secret, bool,
 	switch (label)
 	{
 		case TLS_HKDF_C_HS_TRAFFIC:
+			chunk_clear(&this->handshake_traffic_secrets.client);
+			this->handshake_traffic_secrets.client = chunk_clone(okm);
+			/* otherwise fall through to next state */
```
Comment (tobiasbrunner) [inline src/libtls/tls_hkdf.c]:
```suggestion
			/* fall-through */
```
That's usually enough if no explicit explanation for the fall-through is needed.

File: src/libtls/tls_hkdf.c [hunk 25]
```diff
@@ -501,25 +515,34 @@ static bool get_shared_label_keys(private_tls_hkdf_t *this, chunk_t label,
...
 
+/**
+ * Clean up secrets
+ */
+static void destroy_secrets(cached_secrets_t *secrets)
+{
+	chunk_clear(&secrets->client);
+	chunk_clear(&secrets->server);
+}
+
```
Comment (tobiasbrunner) [inline src/libtls/tls_hkdf.c]: I'd move this either near the struct declaration, or closer to where it's actually used (e.g. right above the `destroy` method).

File: src/libtls/tls_crypto.c [hunk 26]
```diff
@@ -2555,3 +2555,34 @@ enumerator_t *tls_get_supported_key_types(tls_version_t min_version,
 	return enumerator_create_filter(ht->create_enumerator(ht),
 									filter_key_types, ht, destroy_key_types);
 }
+
```
Comment (tobiasbrunner) [inline src/libtls/tls_crypto.c]: As discussed, since the function does not depend on anything in `tls_crypto.c`, it might be preferable to simply rename the function and make it non-static so it can be called from the other file via forward declaration (similar to `tls_write_key_share()`). But maybe move it to `tls_server.c` so all forward declarations between these two files are in `tls_peer.c`.

File: src/libtls/tls_server.c [hunk 27]
```diff
@@ -860,43 +880,28 @@ static status_t process_key_exchange(private_tls_server_t *this,
...
-		sig = bio_reader_create(reader->peek(reader));
-		verified = this->crypto->verify_handshake(this->crypto, public, sig);
-		sig->destroy(sig);
-		if (verified)
-		{
-			this->peer_auth->merge(this->peer_auth, auth, FALSE);
-			break;
-		}
-		DBG1(DBG_TLS, "signature verification failed, trying another key");
+		DBG1(DBG_TLS, "no TLS public key found for peer '%Y'", this->peer);
```
Comment (tobiasbrunner) [inline src/libtls/tls_server.c]: I wonder if the log message in the previous code was clearer for users. For one, "TLS public key" is not really a thing, and "no trusted certificate" might describe the issue better. You could perhaps change this similarly for the client in the `tls_find_public_key()` commit.

File: src/libtls/tls_server.c (hunk 21 above)
Comment (tobiasbrunner) [reply src/libtls/tls_server.c:949]: The commit message looks weird, like all text is on a single line. I'd also add an "also" (i.e. "This commit **also** addresses") because it does this in addition to adding mutual auth support for TLS 1.3 (which might actually be something you could add to the first line of the message, because mutual auth is already supported for older versions).

File: src/libtls/tls_server.c [hunk 28]
```diff
@@ -1350,6 +1359,57 @@ static status_t send_certificate_request(private_tls_server_t *this,
...
+
+		write_certificate_authorities(writer);
+	}
+	else
+	{
+		/* certificate request context as described in RFC 8446, section 4.3.2 */
+		writer->write_uint8(writer, 0);
+
+		authorities = bio_writer_create(64);
+		write_certificate_authorities(authorities);
```
Comment (tobiasbrunner) [inline src/libtls/tls_server.c:1429]: I'd move this down, similar to how you did that with the other extension and its `supported` writer.

File: scripts/tls_test.c [hunk 29]
```diff
@@ -38,7 +38,27 @@ static void usage(FILE *out, char *cmd)
...
+
+	fprintf(out, "\n\nOptions:\n");
+	fprintf(out, "  --help                       print help and exit\n");
+	fprintf(out, "  --connect <address>          connect to a server on dns name or ip address\n");
+	fprintf(out, "  --listen <address>           listen on dns name or ip address\n");
+	fprintf(out, "  --port <port>                specify the port to use\n");
+	fprintf(out, "  --cert <file>                certificate to use to establish and verify TLS connection\n");
+	fprintf(out, "  --key <file>                 private key to use to establish and verify TLS connection\n");
+	fprintf(out, "  --cacert <file>              certificate to use to authenticate clients.\n");
+	fprintf(out, "                               This configuration enables required client authentication\n");
```
Comment (tobiasbrunner) [inline scripts/tls_test.c]: As discussed, clarify the use of these arguments (`--cacert` to verify peer, the others to authenticate itself).

File: scripts/tls_test.c [hunk 30]
```diff
@@ -38,7 +38,27 @@ static void usage(FILE *out, char *cmd)
 {
 	fprintf(out, "usage:\n");
 	fprintf(out, "  %s --connect <address> --port <port> [--key <key] [--cert <file>]+ [--times <n>]\n", cmd);
-	fprintf(out, "  %s --listen <address> --port <port> --key <key> [--cert <file>]+ [--times <n>]\n", cmd);
+	fprintf(out, "  %s --listen <address> --port <port> --key <key> [--cert <file>]+ [--cacert <file>]+ [--times <n>]\n", cmd);
+
+	fprintf(out, "\n\nOptions:\n");
```
Comment (tobiasbrunner) [inline scripts/tls_test.c]: `pki` and `swanctl` both don't add a separator between usage lines and the option descriptions (which are indented more, though - but strangely not the same, `pki` indents with 8 spaces, `swanctl` with 11, not sure why, or why both indent the options that much at all, even without separator, less should be enough). I'm not opposed to adding a separator and, in particular, indenting the options less. But maybe make `options:` lower case (to match `usage:`, which is lower case in the other tools too) with only a single empty line before, or use a single empty line as separator and indent the options by 4 spaces (and reduce the space between options and their descriptions a bit).

**Review by tobiasbrunner — CHANGES_REQUESTED** [review]

File: src/libtls/tls_server.c (hunk 28 above)
Comment (ryru) [reply src/libtls/tls_server.c:1429]: Do you mean moving these two lines:

```
	authorities = bio_writer_create(64);
	write_certificate_authorities(authorities);
```

between the two extensions?

I think I moved this on purpose because of your comment a week ago:

> I'd make it the same as the others, i.e. remove CertificateRequest, because we see sending TLS CertificateRequest handshake in the log right afterwards (well, after the DNs for the CAs - adding them before the other extensions might clarify this even further).

Did I misunderstand you?

Comment (tobiasbrunner) [reply src/libtls/tls_server.c:1429]: I see, sorry about that. The code is just a bit strange this ways and the log messages are too (sending cert reqs, followed by the signature_algorithms extension and only then the certificate_authorities extension). I'd switch the order of the two extensions, so creation and use of the writer are closer together and the log would make a bit more sense (maybe if the cert_authorities extension is sent first, actually moving these lines after the "sending extension" log message would also be OK).

Comment (ryru) [reply src/libtls/tls_server.c:1429]: I addressed this comment in the latest commits.

File: scripts/tls_test.c [hunk 31]
```diff
@@ -37,8 +37,27 @@
 static void usage(FILE *out, char *cmd)
 {
 	fprintf(out, "usage:\n");
-	fprintf(out, "  %s --connect <address> --port <port> [--key <key] [--cert <file>]+ [--times <n>]\n", cmd);
-	fprintf(out, "  %s --listen <address> --port <port> --key <key> [--cert <file>]+ [--times <n>]\n", cmd);
+	fprintf(out, "  %s --connect <address> --port <port> [--key <key] [--cert <file>] [--cacert <file>]+ [--times <n>]\n", cmd);
+	fprintf(out, "  %s --listen <address> --port <port> --key <key> --cert <file> [--cacert <file>]+ [--times <n>]\n", cmd);
+	fprintf(out, "\n");
+	fprintf(out, "options:\n");
+	fprintf(out, "    --help                   print help and exit\n");
```
Comment (tobiasbrunner) [inline scripts/tls_test.c]: Not sure, but you might have overlooked the "or" in my comment about this :) With the "options:" separator, using two spaces indentation should be perfectly fine (and even provides two characters more for the description - I might adapt the other tools to this too because using that much indentation there seems wasteful). The four space indentation was a suggestion in case the options were separated by an empty line only, so they can be better distinguished from the usage lines that are indented by two spaces.

**Review by tobiasbrunner — APPROVED** [review]
Just a small change request in `tls_test.c`, otherwise looks good.

**Comment by ryru** [comment]
Thanks for the feedback.
